### PR TITLE
First pass at pulling out common functionality

### PR DESCRIPTION
**Pull request by se3000**
This is mostly stuff that is shared between the `ruby-ethereum` gem, and the `ethereum-tx` gem. For the most part it's a lot of functionality, with matching tests, pulled from `ruby-ethereum`. Other then that, I spent a bunch of time messing around with name spaces, and trying to get a couple gems to work nicely together in one namespace.

I was able to get the `ethereum-tx` gem to work well with base and not clobber any namespaces
The next step I imagined would be for me to make sure the same is possible with `ruby-ethereum`. Once everything is laid out in `base`, I'll send a PR to replace functionality in `ruby-ethereum` with the common functionality in base.

Sorry this is taking a while. I've been releasing to rubygems for testing purposes. Once it becomes stable, we can tie releases to code review. I'm sending a PR now, partly because it seems like some people are looking at the main repo wondering what's going on. Would love feedback on any of this code, and I'm also curious about what else might be worth moving into the `base` gem. Thanks!

### CODE REVIEW

File: lib/ethereum/base.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+module Ethereum
+  module Base
+
+    BYTE_ZERO = "\x00".freeze
+    GTXCOST = 21000       # TX BASE GAS COST
+    GTXDATANONZERO = 68   # TX DATA NON ZERO BYTE GAS COST
+    GTXDATAZERO = 4       # TX DATA ZERO BYTE GAS COST
+    SECP256K1_N = 115792089237316195423570985008687907852837564279074904382605163141518161494337
+    UINT_MAX = 2**256 - 1
```
Comment (janx) [inline lib/ethereum/base.rb]: We should extract these constants (common constants like `BYTE_ZERO`, gas costs with `GTX` prefix, secp256k1 specific ones) into different modules.

If ethereum-base doesn't include secp256k1 functions, I wonder if we can move `SECP256K1_N` to a better place?

**Comment by janx** [comment]
Thanks for making the first move! :beers:

**Comment by se3000** [comment]
👍 The sooner we get moving, the better.

Took a first pass at organizing Moved some of the constants into `Ethereum::Base::Gas` and `Ethereum::Base::Secp256K1`. We could probably fill out `Gas` with a lot more constants. Might be easier that maintaining that list in multiple places.

We could remove SECP256K1 constants all together, but I figured they would be commonly used. They aren't yet, so we can get rid of them until that's the case.

Between the gas operation costs and the secp variables, it seems like we could include a bunch of constants that we assume will be commonly used, or we could wait until they are needed in a couple places. Do you have any preference?

I kind of prefer including extra constants early. Partly because it'd be easier to include them early than it would be to go back after they are defined in multiple places and update a couple gems to use them. I don't feel super strongly about it, but if you're on board I'll add some more to the gem.

**Comment by janx** [comment]
I would always delay the decision if possible. It's easy to introduce unnecessary global dependencies.

**Pull request merged**